client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -120,10 +125,12 @@ public synchronized void setHosts(HttpHost... hosts) {
         if (hosts == null || hosts.length == 0) {
             throw new IllegalArgumentException("hosts must not be null nor empty");
         }
+        authCache.clear();
         Set<HttpHost> httpHosts = new HashSet<>();
         for (HttpHost host : hosts) {
             Objects.requireNonNull(host, "host cannot be null");
             httpHosts.add(host);
+            authCache.put(host, new BasicScheme());
```

Comment:
I do not think we should hardcode the scheme. This makes the assumption that every host uses preemptive basic auth if basic auth credentials are provided. Some requests may not need preemptive authentication. For example, requests could be made with anonymous access, while others from the same client may require authentication. In that scenario you don't want/need preemptive auth but we are forcing it. 

I think this should be opt-in like the default behavior of HttpClient. We can do this by providing the ability to specify a callback in the builder that when given a HttpHost it will provide the appropriate scheme.